tls/s2n_kex.c

```diff
@@ -60,31 +61,45 @@ static int s2n_write_no_extension(const struct s2n_connection *conn, struct s2n_
     return 0;
 }
 
-static int s2n_check_rsa_key(const struct s2n_kex *kex, const struct s2n_connection *conn)
+static int s2n_check_rsa_key(const struct s2n_cipher_suite *cipher_suite, struct s2n_connection *conn)
 {
     return conn->handshake_params.our_chain_and_key != NULL;
 }
 
-static int s2n_check_dhe(const struct s2n_kex *kex, const struct s2n_connection *conn)
+static int s2n_check_dhe(const struct s2n_cipher_suite *cipher_suite, struct s2n_connection *conn)
 {
     return conn->config->dhparams != NULL;
 }
 
-static int s2n_check_ecdhe(const struct s2n_kex *kex, const struct s2n_connection *conn)
+static int s2n_check_ecdhe(const struct s2n_cipher_suite *cipher_suite, struct s2n_connection *conn)
 {
     return conn->secure.server_ecc_params.negotiated_curve != NULL;
 }
 
-static int s2n_check_kem(const struct s2n_kex *kex, const struct s2n_connection *conn)
+static int s2n_check_kem(const struct s2n_cipher_suite *cipher_suite, struct s2n_connection *conn)
 {
-    return conn->secure.s2n_kem_keys.negotiated_kem != NULL;
+    const struct s2n_iana_to_kem *supported_params = NULL;
+    /* If the cipher suite has no supported KEMs return 0 */
+    if (s2n_cipher_suite_to_kem(cipher_suite->iana_value, &supported_params) != 0) { return 0; }
+    if (supported_params->kem_count == 0) { return 0; }
+
+    struct s2n_blob *proposed_kems = &conn->secure.client_pq_kem_extension;
+    /* If the client did not send a PQ KEM extension the server can pick any paramters it wants */
+    if (proposed_kems == NULL || proposed_kems->data == NULL) {
+        conn->secure.s2n_kem_keys.negotiated_kem = supported_params->kems[0];
```

Comment:
I updated them to be more clear but it might be a bad pattern. It might make more sense to going back to storing mutual KEMs for each cipher suite and then in send key picking the correct one like DH params. Or saving the mutual one during extension parsing like ECDH curves. I went this way to avoid having 2+ supported_KEMs in s2n_crypto_parameters.